File: src/main/java/com/zalphion/featurecontrol/source/PreFetchingFeatureSource.java
package com.zalphion.featurecontrol.source;

import com.zalphion.featurecontrol.bundle.FeatureBundle;
import com.zalphion.featurecontrol.lib.result.Result;
import lombok.*;
import lombok.extern.slf4j.Slf4j;
import org.jspecify.annotations.NonNull;

import java.time.Duration;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;

@Slf4j
public class PreFetchingFeatureSource extends FeatureSource implements AutoCloseable {
    private final @lombok.NonNull FeatureSource inner;
    private final @lombok.NonNull Duration retryInterval;
    private final @lombok.NonNull ScheduledExecutorService scheduler;

    // FIXME don't include in lombok
    private final AtomicReference<FeatureBundle> cache = new AtomicReference<>();

    public PreFetchingFeatureSource(@NonNull @lombok.NonNull FeatureSource inner) {
        this(inner, Duration.ofSeconds(10), Duration.ofSeconds(1), Executors.newSingleThreadScheduledExecutor());
    }

    public PreFetchingFeatureSource(
            @NonNull @lombok.NonNull FeatureSource inner,
            @NonNull @lombok.NonNull Duration refreshInterval,
            @NonNull @lombok.NonNull Duration retryInterval,
            @NonNull @lombok.NonNull ScheduledExecutorService scheduler
    ) {
        this.inner = inner;
        this.retryInterval = retryInterval;
        this.scheduler = Executors.newSingleThreadScheduledExecutor();

        scheduler.scheduleWithFixedDelay(this::fetchNow, 0, refreshInterval.toMillis(), TimeUnit.MILLISECONDS);
    }

    @Override
    protected @NonNull @lombok.NonNull Result<FeatureBundle> getInternal() {
        return Result.successOr(cache.get(), () -> "A bundle has not yet been successfully fetched");
    }

    @Override
    public void close() {
        scheduler.shutdown();
    }

    private void fetchNow() {
        if (scheduler.isShutdown()) return;

        try {
            inner.get().peek(value -> {
                val previous = cache.getAndSet(value);
                if (previous == null) {
                    log.debug("Ready");
                } else {
                    log.trace("Refreshed");
                }
            }).peekFailure(message -> {
                    log.warn(message);
                    if (cache.get() == null && !scheduler.isShutdown()) {
                        scheduler.schedule(this::fetchNow, retryInterval.toMillis(), TimeUnit.MILLISECONDS);
                    }
            });
        } catch (RuntimeException e) {
            log.warn("Error refreshing SDK Bundle", e);
        }
    }
}

File: src/main/kotlin/com/zalphion/featurecontrol/source/PreFetchingFeatureSource.kt
package com.zalphion.featurecontrol.source

import com.zalphion.featurecontrol.Features
import dev.forkhandles.result4k.asFailure
import dev.forkhandles.result4k.asResultOr
import dev.forkhandles.result4k.peek
import dev.forkhandles.result4k.peekFailure
import org.slf4j.LoggerFactory
import java.time.Duration
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicReference

/**
 * Caches a FeatureBundle asynchronously with the help of a background thread.
 * The caller will never be blocked by a network call.
 */
fun FeatureSource.preFetching(
    refreshInternal: Duration = Duration.ofMinutes(1),
    retryInterval: Duration = Duration.ofSeconds(1),
    scheduler: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor(),
) = object: CloseableFeatureSource {
    private val logger = LoggerFactory.getLogger(FeatureSource::class.java)
    private val inner = this@preFetching
    private val cache = AtomicReference<Features>()
    private val firstFetchFuture = CompletableFuture<FeatureSource>()

    private fun fetchNow() {
        if (scheduler.isShutdown) return

        try {
            inner.get().peek {
                cache.set(it)
                if (firstFetchFuture.complete(this)) {
                    logger.debug("Feature Bundle initialized successfully")
                } else {
                    logger.trace("Refreshed SDK Bundle")
                }
            }.peekFailure {
                logger.error("Failed to refresh SDK Bundle: $it")
                if (!firstFetchFuture.isDone && !scheduler.isShutdown) {
                    scheduler.schedule(::fetchNow, retryInterval.toMillis(), TimeUnit.MILLISECONDS)
                }
            }
        } catch (_: InterruptedException) {
            logger.debug("SDK Bundle refresh interrupted")
            Thread.currentThread().interrupt()
        } catch (e: Exception) {
            logger.error("Error refreshing SDK Bundle", e)
        }
    }

    init {
        scheduler.scheduleWithFixedDelay(::fetchNow, 0, refreshInternal.toMillis(), TimeUnit.MILLISECONDS)
    }

    override fun get() = cache.get().asResultOr { "A bundle has not yet been successfully fetched" }

    override fun close() = scheduler.shutdown()
}